# College Major Analysis based on economical factors

## The goal of the analysis is to provide a data model that can help students and parents choose a college major and understand how big a financial difference it makes.

In [1]:
import pandas as pd
from matplotlib import pyplot as plt
import seaborn as sns
import numpy as np
import warnings
warnings.simplefilter(action="ignore", category=FutureWarning)

In [2]:
def load_data(filename):
    """Given a filename of a csv load data into a Pandas dataframe.

        filename - string

        return Pandas dataframe
    """
    return pd.read_csv(filename, encoding='utf-8')

**The dataset files used for this analysis are:**
1. all-ages.csv
2. majors-list.csv
3. women-stem.csv
4. recent-grads.csv
5. grad-students.csv

## DataSet Information
**Some columns used in the dataset are:**

* Rank - Rank by median earnings (the dataset is ordered by this column).
* Major_code - Major code.
* Major - Major description.
* Major_category - Category of major.
* Total - Total number of people with major.
* Sample_size - Sample size (unweighted) of full-time.
* Men - Male graduates.
* Women - Female graduates.
* ShareWomen - Women as share of total.
* Employed - Number employed.
* Median - Median salary of full-time, year-round workers.
* Low_wage_jobs - Number in low-wage service jobs.
* Full_time - Number employed 35 hours or more.
* Part_time - Number employed less than 35 hours.


In [3]:
all_ages = 'all-ages.csv'
grad_students = 'grad-students.csv'
majors_list = 'majors-list.csv'
women_stem = 'women-stem.csv'
recent_grads = 'recent-grads.csv'

In [4]:
all_ages_df = load_data(all_ages)
recent_grads_df = load_data(recent_grads)
grad_students_df = load_data(grad_students)
majors_list_df = load_data(majors_list)
women_stem_df = load_data(women_stem)

## Data Cleaning (Removing records with missing values)

In [5]:
def remove_missing_values(df, df_name):
    original_data_len = len(df)
    print("Before removing the missing data and null values, the dataset '{}' has {} number of records".format(df_name, original_data_len))
    
    #drop records with missing values
    df.dropna(inplace = True)
    
    #Now again let's check the no. of records we have
    cleaned_data_len = len(df)
    print("After removing the missing data and null values, the dataset '{}' has {} number of records\n".format(df_name, cleaned_data_len))

In [6]:
remove_missing_values(all_ages_df, all_ages)
remove_missing_values(recent_grads_df, recent_grads)
remove_missing_values(grad_students_df, grad_students)
remove_missing_values(majors_list_df, majors_list)
remove_missing_values(women_stem_df, women_stem)

Before removing the missing data and null values, the dataset 'all-ages.csv' has 173 number of records
After removing the missing data and null values, the dataset 'all-ages.csv' has 173 number of records

Before removing the missing data and null values, the dataset 'recent-grads.csv' has 173 number of records
After removing the missing data and null values, the dataset 'recent-grads.csv' has 172 number of records

Before removing the missing data and null values, the dataset 'grad-students.csv' has 173 number of records
After removing the missing data and null values, the dataset 'grad-students.csv' has 173 number of records

Before removing the missing data and null values, the dataset 'majors-list.csv' has 174 number of records
After removing the missing data and null values, the dataset 'majors-list.csv' has 173 number of records

Before removing the missing data and null values, the dataset 'women-stem.csv' has 76 number of records
After removing the missing data and null values,

In [7]:
all_ages_df.head(10)

Major_code                                  Major  \
0        1100                    GENERAL AGRICULTURE   
1        1101  AGRICULTURE PRODUCTION AND MANAGEMENT   
2        1102                 AGRICULTURAL ECONOMICS   
3        1103                        ANIMAL SCIENCES   
4        1104                           FOOD SCIENCE   
5        1105             PLANT SCIENCE AND AGRONOMY   
6        1106                           SOIL SCIENCE   
7        1199              MISCELLANEOUS AGRICULTURE   
8        1301                  ENVIRONMENTAL SCIENCE   
9        1302                               FORESTRY   

                    Major_category   Total  Employed  \
0  Agriculture & Natural Resources  128148     90245   
1  Agriculture & Natural Resources   95326     76865   
2  Agriculture & Natural Resources   33955     26321   
3  Agriculture & Natural Resources  103549     81177   
4  Agriculture & Natural Resources   24280     17281   
5  Agriculture & Natural Resources   79409     63043   
6  Agriculture & Natural Resources    6586      4926   
7  Agriculture & Natural Resources    8549      6392   
8           Biology & Life Science  106106     87602   
9  Agriculture & Natural Resources   69447     48228   

   Employed_full_time_year_round  Unemployed  Unemployment_rate  Median  \
0                          74078        2423           0.026147   50000   
1                          64240        2266           0.028636   54000   
2                          22810         821           0.030248   63000   
3                          64937        3619           0.042679   46000   
4                          12722         894           0.049188   62000   
5                          51077        2070           0.031791   50000   
6                           4042         264           0.050867   63000   
7                           5074         261           0.039230   52000   
8                          65238        4736           0.051290   52000   
9                          39613        2144           0.042563   58000   

   P25th    P75th  
0  34000  80000.0  
1  36000  80000.0  
2  40000  98000.0  
3  30000  72000.0  
4  38500  90000.0  
5  35000  75000.0  
6  39400  88000.0  
7  35000  75000.0  
8  38000  75000.0  
9  40500  80000.0

In [8]:
rows, cols = all_ages_df.shape

print("The number of rows:", rows)
print("The number of columns:", cols)

The number of rows: 173
The number of columns: 11


In [9]:
grad_students_df.head(10)

Major_code                                     Major  \
0        5601                     CONSTRUCTION SERVICES   
1        6004         COMMERCIAL ART AND GRAPHIC DESIGN   
2        6211                    HOSPITALITY MANAGEMENT   
3        2201    COSMETOLOGY SERVICES AND CULINARY ARTS   
4        2001                COMMUNICATION TECHNOLOGIES   
5        3201                           COURT REPORTING   
6        6206          MARKETING AND MARKETING RESEARCH   
7        1101     AGRICULTURE PRODUCTION AND MANAGEMENT   
8        2101  COMPUTER PROGRAMMING AND DATA PROCESSING   
9        1904          ADVERTISING AND PUBLIC RELATIONS   

                        Major_category  Grad_total  Grad_sample_size  \
0  Industrial Arts & Consumer Services        9173               200   
1                                 Arts       53864               882   
2                             Business       24417               437   
3  Industrial Arts & Consumer Services        5411                72   
4              Computers & Mathematics        9109               171   
5                  Law & Public Policy        1542                22   
6                             Business      190996              3738   
7      Agriculture & Natural Resources       17488               386   
8              Computers & Mathematics        5611                98   
9          Communications & Journalism       33928               688   

   Grad_employed  Grad_full_time_year_round  Grad_unemployed  \
0           7098                       6511              681   
1          40492                      29553             2482   
2          18368                      14784             1465   
3           3590                       2701              316   
4           7512                       5622              466   
5           1008                        860                0   
6         151570                     123045             8324   
7          13104                      11207              473   
8           4716                       3981              119   
9          28517                      22523              899   

   Grad_unemployment_rate  Grad_median  ...  Nongrad_total  Nongrad_employed  \
0                0.087543      75000.0  ...          86062             73607   
1                0.057756      60000.0  ...         461977            347166   
2                0.073867      65000.0  ...         179335            145597   
3                0.080901      47000.0  ...          37575             29738   
4                0.058411      57000.0  ...          53819             43163   
5                0.000000      75000.0  ...           8921              6967   
6                0.052059      80000.0  ...        1029181            817906   
7                0.034838      67000.0  ...          89169             71781   
8                0.024612      85000.0  ...          28314             22024   
9                0.030562      60000.0  ...         163435            127832   

   Nongrad_full_time_year_round  Nongrad_unemployed  \
0                         62435                3928   
1                        250596               25484   
2                        113579                7409   
3                         23249                1661   
4                         34231                3389   
5                          6063                 518   
6                        662346               45519   
7                         61335                1869   
8                         18381                2222   
9                        100330                8706   

   Nongrad_unemployment_rate  Nongrad_median  Nongrad_P25  Nongrad_P75  \
0                   0.050661         65000.0        47000      98000.0   
1                   0.068386         48000.0        34000      71000.0   
2                   0.048423         50000.0        35000      75000.0   
3                   0.052900         41600.0        29000      60000.0  

In [10]:
recent_grads_df.head(10)

Rank  Major_code                                      Major    Total  \
0     1        2419                      PETROLEUM ENGINEERING   2339.0   
1     2        2416             MINING AND MINERAL ENGINEERING    756.0   
2     3        2415                  METALLURGICAL ENGINEERING    856.0   
3     4        2417  NAVAL ARCHITECTURE AND MARINE ENGINEERING   1258.0   
4     5        2405                       CHEMICAL ENGINEERING  32260.0   
5     6        2418                        NUCLEAR ENGINEERING   2573.0   
6     7        6202                          ACTUARIAL SCIENCE   3777.0   
7     8        5001                 ASTRONOMY AND ASTROPHYSICS   1792.0   
8     9        2414                     MECHANICAL ENGINEERING  91227.0   
9    10        2408                     ELECTRICAL ENGINEERING  81527.0   

       Men    Women     Major_category  ShareWomen  Sample_size  Employed  \
0   2057.0    282.0        Engineering    0.120564           36      1976   
1    679.0     77.0        Engineering    0.101852            7       640   
2    725.0    131.0        Engineering    0.153037            3       648   
3   1123.0    135.0        Engineering    0.107313           16       758   
4  21239.0  11021.0        Engineering    0.341631          289     25694   
5   2200.0    373.0        Engineering    0.144967           17      1857   
6   2110.0   1667.0           Business    0.441356           51      2912   
7    832.0    960.0  Physical Sciences    0.535714           10      1526   
8  80320.0  10907.0        Engineering    0.119559         1029     76442   
9  65511.0  16016.0        Engineering    0.196450          631     61928   

   ...  Part_time  Full_time_year_round  Unemployed  Unemployment_rate  \
0  ...        270                  1207          37           0.018381   
1  ...        170                   388          85           0.117241   
2  ...        133                   340          16           0.024096   
3  ...        150                   692          40           0.050125   
4  ...       5180                 16697        1672           0.061098   
5  ...        264                  1449         400           0.177226   
6  ...        296                  2482         308           0.095652   
7  ...        553                   827          33           0.021167   
8  ...      13101                 54639        4650           0.057342   
9  ...      12695                 41413        3895           0.059174   

   Median  P25th   P75th  College_jobs  Non_college_jobs  Low_wage_jobs  
0  110000  95000  125000          1534               364            193  
1   75000  55000   90000           350               257             50  
2   73000  50000  105000           456               176              0  
3   70000  43000   80000           529               102              0  
4   65000  50000   75000         18314              4440            972  
5   65000  50000  102000          1142               657            244  
6   62000  53000   72000          1768               314            259  
7   62000  31500  109000           972               500            220  
8   60000  48000   70000         52844             16384           3253  
9   60000  45000   72000         45829             10874           3170  

[10 rows x 21 columns]

In [11]:
majors_list_df.head(10)

FOD1P                                  Major  \
0  1100                    GENERAL AGRICULTURE   
1  1101  AGRICULTURE PRODUCTION AND MANAGEMENT   
2  1102                 AGRICULTURAL ECONOMICS   
3  1103                        ANIMAL SCIENCES   
4  1104                           FOOD SCIENCE   
5  1105             PLANT SCIENCE AND AGRONOMY   
6  1106                           SOIL SCIENCE   
7  1199              MISCELLANEOUS AGRICULTURE   
8  1302                               FORESTRY   
9  1303           NATURAL RESOURCES MANAGEMENT   

                    Major_Category  
0  Agriculture & Natural Resources  
1  Agriculture & Natural Resources  
2  Agriculture & Natural Resources  
3  Agriculture & Natural Resources  
4  Agriculture & Natural Resources  
5  Agriculture & Natural Resources  
6  Agriculture & Natural Resources  
7  Agriculture & Natural Resources  
8  Agriculture & Natural Resources  
9  Agriculture & Natural Resources

In [12]:
women_stem_df.head(10)

Rank  Major_code                                      Major  \
0     1        2419                      PETROLEUM ENGINEERING   
1     2        2416             MINING AND MINERAL ENGINEERING   
2     3        2415                  METALLURGICAL ENGINEERING   
3     4        2417  NAVAL ARCHITECTURE AND MARINE ENGINEERING   
4     5        2418                        NUCLEAR ENGINEERING   
5     6        2405                       CHEMICAL ENGINEERING   
6     7        5001                 ASTRONOMY AND ASTROPHYSICS   
7     8        2414                     MECHANICAL ENGINEERING   
8     9        2401                      AEROSPACE ENGINEERING   
9    10        2408                     ELECTRICAL ENGINEERING   

      Major_category  Total    Men  Women  ShareWomen  Median  
0        Engineering   2339   2057    282    0.120564  110000  
1        Engineering    756    679     77    0.101852   75000  
2        Engineering    856    725    131    0.153037   73000  
3        Engineering   1258   1123    135    0.107313   70000  
4        Engineering   2573   2200    373    0.144967   65000  
5        Engineering  32260  21239  11021    0.341631   65000  
6  Physical Sciences   1792    832    960    0.535714   62000  
7        Engineering  91227  80320  10907    0.119559   60000  
8        Engineering  15058  12953   2105    0.139793   60000  
9        Engineering  81527  65511  16016    0.196450   60000

# Summary statistics:

**recent-grads.csv**

In [13]:
recent_grads_df.describe()

Rank   Major_code          Total            Men          Women  \
count  172.000000   172.000000     172.000000     172.000000     172.000000   
mean    87.377907  3895.953488   39370.081395   16723.406977   22646.674419   
std     49.983181  1679.240095   63483.491009   28122.433474   41057.330740   
min      1.000000  1100.000000     124.000000     119.000000       0.000000   
25%     44.750000  2403.750000    4549.750000    2177.500000    1778.250000   
50%     87.500000  3608.500000   15104.000000    5434.000000    8386.500000   
75%    130.250000  5503.250000   38909.750000   14631.000000   22553.750000   
max    173.000000  6403.000000  393735.000000  173809.000000  307087.000000   

       ShareWomen  Sample_size      Employed      Full_time      Part_time  \
count  172.000000   172.000000     172.00000     172.000000     172.000000   
mean     0.522223   357.941860   31355.80814   26165.767442    8877.232558   
std      0.231205   619.680419   50777.42865   42957.122320   14679.038729   
min      0.000000     2.000000       0.00000     111.000000       0.000000   
25%      0.336026    42.000000    3734.75000    3181.000000    1013.750000   
50%      0.534024   131.000000   12031.50000   10073.500000    3332.500000   
75%      0.703299   339.000000   31701.25000   25447.250000    9981.000000   
max      0.968954  4212.000000  307933.00000  251540.000000  115172.000000   

       Full_time_year_round    Unemployed  Unemployment_rate         Median  \
count            172.000000    172.000000         172.000000     172.000000   
mean           19798.843023   2428.412791           0.068024   40076.744186   
std            33229.227514   4121.730452           0.030340   11461.388773   
min              111.000000      0.000000           0.000000   22000.000000   
25%             2474.750000    299.500000           0.050261   33000.000000   
50%             7436.500000    905.000000           0.067544   36000.000000   
75%            17674.750000   2397.000000           0.087247   45000.000000   
max           199897.000000  28169.000000           0.177226  110000.000000   

              P25th          P75th   College_jobs  Non_college_jobs  \
count    172.000000     172.000000     172.000000        172.000000   
mean   29486.918605   51386.627907   12387.401163      13354.325581   
std     9190.769927   14882.278650   21344.967522      23841.326605   
min    18500.000000   22000.000000       0.000000          0.000000   
25%    24000.000000   41750.000000    1744.750000       1594.000000   
50%    27000.000000   47000.000000    4467.500000       4603.500000   
75%    33250.000000   58500.000000   14595.750000      11791.750000   
max    95000.000000  125000.000000  151643.000000     148395.000000   

       Low_wage_jobs  
count     172.000000  
mean     3878.633721  
std      6960.467621  
min         0.000000  
25%       336.750000  
50%      1238.500000  
75%      3496.000000  
max     48207.000000

**majors-list.csv**

In [14]:
majors_list_df.describe()

FOD1P                                     Major Major_Category
count    173                                       173            173
unique   173                                       173             16
top     2312  TRANSPORTATION SCIENCES AND TECHNOLOGIES    Engineering
freq       1                                         1             29

**all-ages.csv**

In [15]:
all_ages_df.describe()

Major_code         Total      Employed  Employed_full_time_year_round  \
count   173.000000  1.730000e+02  1.730000e+02                   1.730000e+02   
mean   3879.815029  2.302566e+05  1.661620e+05                   1.263078e+05   
std    1687.753140  4.220685e+05  3.073244e+05                   2.424254e+05   
min    1100.000000  2.396000e+03  1.492000e+03                   1.093000e+03   
25%    2403.000000  2.428000e+04  1.728100e+04                   1.272200e+04   
50%    3608.000000  7.579100e+04  5.656400e+04                   3.961300e+04   
75%    5503.000000  2.057630e+05  1.428790e+05                   1.110250e+05   
max    6403.000000  3.123510e+06  2.354398e+06                   1.939384e+06   

          Unemployed  Unemployment_rate         Median         P25th  \
count     173.000000         173.000000     173.000000    173.000000   
mean     9725.034682           0.057355   56816.184971  38697.109827   
std     18022.040192           0.019177   14706.226865   9414.524761   
min         0.000000           0.000000   35000.000000  24900.000000   
25%      1101.000000           0.046261   46000.000000  32000.000000   
50%      3619.000000           0.054719   53000.000000  36000.000000   
75%      8862.000000           0.069043   65000.000000  42000.000000   
max    147261.000000           0.156147  125000.000000  78000.000000   

               P75th  
count     173.000000  
mean    82506.358382  
std     20805.330126  
min     45800.000000  
25%     70000.000000  
50%     80000.000000  
75%     95000.000000  
max    210000.000000

**women-stem.csv**

In [ ]:
women_stem_df.describe()

**grad-students.csv**

In [ ]:
grad_students_df.describe()

# Analysis and Visualization

**grad-students.csv**

In [ ]:
# Plot Correlation matrix
def plot_CorrMatrix(df, csv_name):
    corr = df.corr()
    corr.style.background_gradient(cmap='coolwarm')
    fig, ax = plt.subplots(figsize=(8, 8))
    ax.matshow(corr)
    plt.xticks(range(len(corr.columns)), corr.columns, rotation=90)
    plt.yticks(range(len(corr.columns)), corr.columns)
    plt.gca().xaxis.tick_bottom()  # to make sure label appears at the bottom
    plt.title(f'Correlation Matrix for {csv_name}', fontsize=18)
    
    plt.show()

In [ ]:
plot_CorrMatrix(grad_students_df, grad_students)

**recent-grads.csv**

In [ ]:
plot_CorrMatrix(recent_grads_df, recent_grads)

# Majors with median salary above $60,000

In [ ]:
high_median_salary = recent_grads_df[recent_grads_df["Median"] > 60000].sort_values("Median")
high_median_salary.plot(x="Major", y=["P25th", "Median", "P75th"], kind="bar", figsize=(8,8))

plt.show()

### From the above bar chart, it is evident that the Petroleum Engineering major graduates got paid the highest out all other majors with a median salary of over $100,000. 

In [ ]:
recent_grads_df.plot(x="Rank", y=["P25th", "Median", "P75th"], figsize=(8,8))
plt.show()

### From the above plot, we can make the following observations:
1. **Median income decreases as the rank decreases since the rank determines the median income**
2. **Some of the majors in the figure have large gaps between the 25th and 75th percentiles. This shows that graduates  with those degrees may earn notably lower or notably higher than the mdedian income.**
3. **All remaining majors have very small gaps between the 25th and 75th percentiles which means that the salary of the graduates with those degrees is very close to the median income.**

## Creating Scatter plots to analyze the relationships between different columns in the recent_grads dataset

In [ ]:
# Total v/ Employed

# to exclude the tail of the series from analysis
recent_grads_modified=recent_grads_df[recent_grads_df["Total"]< recent_grads_df["Total"].quantile(0.9)]

recent_grads_modified.plot(x='Total', y='Employed', kind='scatter', title="Total v/s Employed (reduced dataset)", figsize=(8,8))
plt.show()

### The above scatter plot between Total v/s Employed shows a strong positive correlation. However, we should not forget that correlation doesn't imply causation.

In [ ]:
#Sample_size v/s Median

recent_grads_df.plot(x='Sample_size', y='Median', kind='scatter', title='Sample_size v/s Median', figsize=(8,8))
plt.show()

### From the above scatter plot, we can see that:
1. **Median salaries of the full time employees varies greatly in the lower tier of the sample size**
2. **We can't see any linear relation between the sample size and the median salary**

### Scatter plot with Median and ShareWomen

In [ ]:
plt.figure(figsize=(10,10))
plt.ylim(0,100000)

sns.regplot(x='ShareWomen', y='Median', data=recent_grads_df, scatter=False)
sns.scatterplot(x='ShareWomen', y='Median', data=recent_grads_df, hue='Major_category',
                size='Sample_size', sizes=(10, 800), alpha=0.5, palette='bright')

plt.show()

### From the above Scatterplot, we can see the following observations:
* Slightly but surely we can see a negative correlation between Median and ShareWomen
* The health sector seems like a clear outlier as it has a high ShareWomen along with a high Median salary.
* Petroleum Engineering despite of a small sample size has the highest Median salary.

# Data Distribution

In [ ]:
columns = ["Sample_size", "Median", "Employed", "Full_time", "ShareWomen", "Unemployment_rate", "Men", "Women"]

figure = plt.figure(figsize=(7,25))
for i in range(0,4):
    ax = figure.add_subplot(4,1,i+1)
    ax.set_xlabel(columns[i])
    ax = recent_grads_df[columns[i]].plot(kind='hist', rot=30)
    
figure = plt.figure(figsize=(7,25))
for i in range(4,8):
    ax = figure.add_subplot(4,1,i-3)
    ax.set_xlabel(columns[i])
    ax = recent_grads_df[columns[i]].plot(kind='hist', rot=30)

### From the above histograms, we can conclude that:
1. **From the sample size histogram, it is clearly visible that the majority of the sample size of full-time workers is notably less than 500**
2. **From the Median salary histogram, we can see that the median salary range of recent graduates lies between \\$30,000- \\$40,000**
3. **From the employed column histogram, we can see that less than 50,000 graduate students were able to find some kind of employment** 
4. **From the Full_time histogram, we can see that number of full time employees is less than 25,000**
5. **From the ShareWomen histogram, we can see that share of the women in the total lies in the range of 0.2%-0.75%**
6. **We can see that the unemployment rate has never exceeded over 17.7% from the unemployment_rate histogram**
7. **The last 2 histograms which are number of men and number of women shows us that the number of men is notably similar to the number of women**

# Bar Plots

## Unemployment rate for top 10 ranking majors based on median salary

In [ ]:
# slicing the list up to 10 will give us top 10 ranked majors
recent_grads_df[:10].plot.bar(x='Major', y='Unemployment_rate', figsize=(8,8))

plt.title("Unemployment_rate based on Major")
plt.ylabel("Unemployment_rate")

plt.show()

### From the above Bar plot, we can see that unemployment rate for the top 10 ranked majors is notably low. However, Nuclear Engineering looks like a slight outlier as it has a significantly higher unemployment rate compared to other top 10 ranked majors based on the median salary.

## Unemployment rate for the 10 least ranked majors based on median salary

In [ ]:
# slicing the list from 163 to the end since there are 173 total majors will give us least 10 ranked majors
recent_grads_df[163:].plot.bar(x='Major', y='Unemployment_rate', figsize=(8,8))

plt.title("Unemployment Rate based on Major")
plt.ylabel("Unemployment_rate")
plt.show()

### From the above Bar plot, we can also see that out of 10 least ranked majors, Clinical Psychology looks has the highest unemployment rate followed by Library Science and Other Foreign Languages.

## ShareWomen(women as share of total) for the top 10 ranked majors based on median salary

In [ ]:
# slicing the list up to 10 will give us the top 10 ranked majors
recent_grads_df[:10].plot.bar(x='Major', y='ShareWomen', figsize=(8,8))

plt.title("ShareWomen based on Major")
plt.ylabel("ShareWomen")
plt.show()

### From the above Bar plot, we can see that the top 10 ranking Majors based on Median Salary are notably dominated by males. It is also evident that 7 out of the 10 top ranked Majors has less than 20% female graduates. Astronomy and Astrophysics has notably high womenshare out all other top 10 ranked majors.

## ShareWomen(women as share of total) for the 10 least ranked majors based on median salary

In [ ]:
# slicing the list from 163 to the end since there are 173 total majors will give us least 10 ranked majors
recent_grads_df[163:].plot.bar(x='Major', y='ShareWomen', figsize=(8,8))

plt.title("ShareWomen by Major")
plt.ylabel("ShareWomen")
plt.show()

### From the above Bar plot, we can clearly see that the 10 least ranked Majors are dominated by Female graduates. There are more than 60% female graduates in the 10 least ranked majors based on the median salary.

# Top 10 Popular Majors

In [ ]:
# ascending is set to False as we want top 10 
top_10_popular =recent_grads_df.sort_values(by="Total",axis=0,ascending=False)

fig, ax = plt.subplots(figsize=(8,8))

# barh is used to to make a horizontal bar plot
ax.barh(top_10_popular["Major"][:10],top_10_popular["Total"][:10] )
# slicing all the list to 10 since we want top 10 majors
ax.set_yticks(top_10_popular["Major"][:10])
ax.set_yticklabels(top_10_popular["Major"][:10])
ax.invert_yaxis()  # to read the labels from top-to-bottom
ax.set_xlabel('Total Students')
ax.set_title('Top 10 Majors')

plt.show()

### Hence, from the above plot we can conclude that Psychology is the most popular major out of all other majors.  Political Science and Government is the last major on the top 10 popular majors as it barely made to the top 10 popular majors list.

# 10 Least Popular Majors

In [ ]:
# ascending is set to true to get the least 10 popular majors
least_10_popular = recent_grads_df.sort_values(by="Total",axis=0,ascending=True)

fig, ax = plt.subplots(figsize=(8,8))

# barh to make a horizontal bar plot
ax.barh(least_10_popular["Major"][:10],least_10_popular["Total"][:10])
ax.set_yticks(least_10_popular["Major"][:10])
ax.set_yticklabels(least_10_popular["Major"][:10])
ax.invert_yaxis()  # labels read top-to-bottom
ax.set_xlabel('Total Students')
ax.set_title('Least 10 popular Majos')
plt.show()

### Hence, from the above plot we can conclude that Military Technologies is the least popular major out of all other majors and Court Reporting is at the bottom of top 10 least popular majors

# 15 Most Popular majors breakdown by gender

In [ ]:
# head is set to 15 to get top 15 majors
graduates_major_grouping = recent_grads_df.groupby('Major').sum().sort_values('Total', ascending=False).head(15).reset_index()

major_gender_df = graduates_major_grouping.sort_values('Total')[['Major', 'Men', 'Women']]
major_gender_df.set_index('Major', inplace=True)
major_gender_df.plot(kind='barh', stacked=True, figsize=(8,8), width=0.6)

plt.xlabel('Total')
plt.title('Top 15 most popular majors with gender breakdown', fontdict={'fontsize': 20})
plt.show()

### From the above plot, we can see that:
* Psychology seems to be the most popular major which is highly skewed towards women
* Business Management & Administration and General Business are split evenly among both men and women
* Computer Science seems to be highly skewed towards men
* Nursing, Elementary Education and General Education are the most skewed towards Women out of all the majors

# Median salary variation in each Major category

In [ ]:
plt.figure(figsize=(8,8))
sns.set_style('whitegrid')

variation = recent_grads_df.loc[:, ['Major_category', 'Median']].groupby('Major_category').median() .sort_values(by='Median', ascending=False)

sns.boxplot(x='Median', y='Major_category', data=recent_grads_df, order=variation.index)
plt.show()

### From the above plot, we can see that:
1. **Out of all the major categories, STEM Majors along with Business majors are the highest paid majors.**
2. **Engineering is at the top of the list and the engineering graduates have a salary which varies between \\$40,000 and just below \\$78,000 with the median salary at just below \\$60,000**
3. **We can also see that most women prefer to go in the Engineering field as it is a less favorable major category by the Women**

# Methods

## 1. Clustering (DBSCAN)

In [ ]:
from sklearn import preprocessing

grad_students_clone = grad_students_df.copy()
grad_students_clone["employment_rate"] = grad_students_clone["Grad_employed"]/(grad_students_clone["Grad_employed"]+grad_students_clone["Grad_unemployed"])

# Label encoding major category values
le = preprocessing.LabelEncoder()
grad_students_clone["Major_category_labelled"] = grad_students_clone[["Major_category"]].apply(le.fit_transform)

In [ ]:
def print_labelled_categories(df):
    dict = {}
    for index, row in df.iterrows():
        if row["Major_category_labelled"] not in dict:
            dict[row["Major_category_labelled"]] = row["Major_category"]
    
    print("Mapping of Major_category\n")
    for key in sorted(dict):
        print("%s: %s" % (key, dict[key]))

print_labelled_categories(grad_students_clone)

In [ ]:
from sklearn.cluster import DBSCAN
# from sklearn.preprocessing import StandardScaler

# Compute DBSCAN
def dbscan(X, eps, min_samples):
    # X = StandardScaler().fit_transform(X)
    db = DBSCAN(eps=eps, min_samples=min_samples)
    db.fit(X)
    labels = db.labels_
    n_clusters_ = len(set(labels))
    y_pred = db.fit_predict(X)
    plt.scatter(X[:,0], X[:,1],c=y_pred, cmap='Paired')
    plt.title("Estimated number of clusters: {}".format(n_clusters_))

In [ ]:
transformed_grad_1 = grad_students_clone[["Grad_median","Major_category_labelled"]].copy()
dbscan(transformed_grad_1.values, 2000, 4)

In [ ]:
transformed_grad_2 = grad_students_clone[["employment_rate","Major_category_labelled"]].copy()
dbscan(transformed_grad_2.values, 0.005, 4)

### From the above two graphs, we can see that:
1. **Most of the majors are likely to pay graduate students between \\$60,000-\\$80,000 except category 1 and 6 (Arts, Education)**
2. **Employment rates are pretty high in category 6, but it is infeasible considering the median pay**
3. **From graph 2 we can decipher that even though certain majors fall under the same category, the employment opprtunities vary vastly (category 1,5,10)**
4. **Category 13 (Physical Sciences) seems to be the safest choice with good pay and employment opportunities. But if one is willing to aim for a high end salary, category 7 (Engineering) is a good choice as it pays better than every other category, but some majors have lower employment than others**

## 2. Random Forest Regression

In [ ]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split

In [ ]:
recent_grads_df=pd.read_csv("recent-grads.csv")
print(recent_grads_df.shape)
print(recent_grads_df.columns)

In [ ]:
data_x=recent_grads_df[['Major_category',"Unemployment_rate",'College_jobs', 'Non_college_jobs','Low_wage_jobs']]

In [ ]:
print(data_x.shape)

In [ ]:
#Applying one hot encoding on major category
onenc=OneHotEncoder()
x_enc=onenc.fit_transform(data_x.iloc[:,0:1]).toarray()
x_enc =pd.DataFrame(x_enc)

In [ ]:
print(x_enc.head())

In [ ]:
df2=pd.concat([x_enc,data_x.iloc[:,1:]],axis=1)

In [ ]:
print(df2.head)

In [ ]:
x=df2.iloc[:,:-1]
y=df2.iloc[:,[-1]]

In [ ]:
print(x.shape)
print(y.shape)

In [ ]:
#Splitting dataset in training dataset and testing dataset
xtrain,xtest,ytrain,ytest=train_test_split(x,y,train_size=0.9,random_state=42)
print(xtrain.shape)
print(ytrain.shape)
print(xtrain.head())

In [ ]:
#Creating a Random Forest Regression model and fitting it to the training data
forest=RandomForestRegressor(n_estimators=1400,random_state=42)
forest.fit(xtrain,ytrain)

#Predicting values for test data
ypred=forest.predict(xtest)
testdf=pd.DataFrame()
print("Model training executed successfully")

In [ ]:
print(ytest)
print(ypred)


In [ ]:
%matplotlib inline

from sklearn.datasets import load_diabetes

import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA, TruncatedSVD
from sklearn.manifold import LocallyLinearEmbedding, TSNE, Isomap
from time import time
from matplotlib.ticker import NullFormatter
import umap.umap_ as umap
from functools import reduce

plt.style.use('ggplot')
#To suppress unnecesary future warings
from warnings import simplefilter
simplefilter(action='ignore', category=FutureWarning)

In [ ]:
df_all_ages_cpy = all_ages_df.copy()
df_grad_students_cpy = grad_students_df.copy()
df_women_stem_cpy = women_stem_df.copy()
df_recent_grads_cpy = recent_grads_df.copy()



df_women_stem_cpy=df_women_stem_cpy.rename(columns={"Median":"W-stem_Median","Total":"W-stem_Total","Men": "W-stem_Men",
                        "Women": "W-stem_Women"})
df_recent_grads_cpy=df_recent_grads_cpy.rename(columns={"Total":"Rec_grad_Total","Men": "Rec_grad_Men", "Women": "Rec_grad_Women",
                   "ShareWomen":"Rec_grad_ShareWomen","Employed":"Rec_grad_Employed",
                   "Unemployed":"Rec_grad_Unemployed","Unemployment_rate":"Rec_grad_Unemployment_rate",
                  "Median":"Rec_grad_Median","P25th":"Rec_grad_P25th","P75th":"Rec_grad_P75th"})


dfs=[df_all_ages_cpy,df_grad_students_cpy,df_women_stem_cpy,df_recent_grads_cpy]
df= reduce(lambda left,right: pd.merge(left,right,on=['Major_code','Major','Major_category']), dfs)


df=pd.get_dummies(df,columns=['Major','Major_category','Major_code'])
df.fillna(value=0, inplace=True)
X = df.values #feature matrix

In [ ]:
df.shape
df.head()

# **1) PCA - Principal Component Analysis**

In [ ]:
#create the object
scaler =  StandardScaler()

#calculate the mean and standard deviation
scaler.fit(X)

#transform the values
X_scaled = scaler.transform(X)

In [ ]:
pca = PCA(n_components= .95 ,random_state = 2020)
pca.fit(X_scaled)
X_pca = pca.transform(X_scaled)

In [ ]:
print('Variance of all = ', sum(pca.explained_variance_ratio_*100))

In [ ]:
plt.plot(pca.explained_variance_ratio_)
plt.xlabel('Number of Components')
plt.ylabel('Explained Variance')

In [ ]:
print('Variance for Principal Components = ', np.cumsum(pca.explained_variance_ratio_ * 100)[69])

# **2) SVD - Single Value Decomposition**

In [ ]:
#Fitting the data in SVD class with the features
trun_svd =  TruncatedSVD(n_components=163)
svd = trun_svd.fit_transform(X)

In [ ]:
print('Variance explained by th Principal Components = ', np.cumsum(trun_svd.explained_variance_ratio_ * 100)[1])

In [ ]:
trun_svd =  TruncatedSVD()
A_transformed = trun_svd.fit_transform(df)

In [ ]:
print("Transformed Matrix:")
print(A_transformed)

In [ ]:
print('Variance of all features = ', sum(trun_svd.explained_variance_ratio_*100))

In [ ]:
plt.plot(np.cumsum(trun_svd.explained_variance_ratio_))
plt.xlabel('Number of Components')
plt.ylabel('Explained Variance')

# **3) LLE - Locally Linear Embedding**

In [ ]:
embedding = LocallyLinearEmbedding(n_components = 70)
X_transformed = embedding.fit_transform(X)
X_transformed.shape

In [ ]:
print(X_transformed)

In [ ]:
plt.style.use('seaborn-whitegrid')
plt.figure(figsize = (10, 6))
c_map = plt.cm.get_cmap('jet', 2)
plt.scatter(X_transformed[:, 0], X_transformed[:, 1], s = 15,
            cmap = c_map)
plt.colorbar()
plt.xlabel('LLE-1') , plt.ylabel('LLE-2')
plt.show()

# **4) tSNE - t Distributed Stochastic Neighbor Embedding**

In [ ]:
X_embedded = TSNE().fit_transform(df)
X_embedded.shape

In [ ]:
plt.style.use('seaborn-whitegrid')
plt.figure(figsize = (10, 6))
c_map = plt.cm.get_cmap('jet', 2)
plt.scatter(X_embedded[:, 0], X_embedded[:, 1], s = 15,
            cmap = c_map )
plt.colorbar()
plt.xlabel('tSNE-1') , plt.ylabel('tSNE-2')
plt.show()

# **5) ISOMAP - Isometric Mapping**

In [ ]:
embedding = Isomap()
X_transformed = embedding.fit_transform(X)
X_transformed.shape

In [ ]:
plt.style.use('seaborn-whitegrid')
plt.figure(figsize = (10, 6))
c_map = plt.cm.get_cmap('jet', 2)
plt.scatter(X_transformed[:, 0], X_transformed[:, 1], s = 15,
            cmap = c_map )
plt.colorbar()
plt.xlabel('ISOMAP-1') , plt.ylabel('ISOMAP-2')
plt.show()

# **6) UMAP**

In [ ]:
trans = umap.UMAP(n_neighbors=5).fit(df)

In [ ]:
plt.scatter(trans.embedding_[:, 0], trans.embedding_[:, 1], s = 5, cmap = 'Spectral')
plt.title('Embedding by UMAP', fontsize=24);

In [ ]:
plt.style.use('seaborn-whitegrid')
plt.figure(figsize = (10, 6))
c_map = plt.cm.get_cmap('jet', 2)
plt.scatter(trans.embedding_[:, 0], trans.embedding_[:, 1], s = 15,
            cmap = c_map )
plt.colorbar()
plt.xlabel('UMAP-1') , plt.ylabel('UMAP-2')
plt.show()

# **Plotting all Methods** (For side by side comparison)

In [ ]:
n_neighbors = 10
# Creating the plot
fig = plt.figure(figsize=(20, 20))
fig.suptitle("Manifold Learning with %i points, %i neighbors"
             % (1000, n_neighbors), fontsize=14)

method_keys = ["LLE","t-SNE","UMAP","ISOMAP","SVD","PCA"]
methods = {}
methods['LLE'] = LocallyLinearEmbedding(n_neighbors=n_neighbors)
methods['t-SNE'] = TSNE(init='pca')
methods['UMAP'] = umap.UMAP()
methods['ISOMAP'] = Isomap()
methods['SVD'] = TruncatedSVD()
methods['PCA']=PCA()
# Plotting the results
for i,method in enumerate(method_keys):
    t0 = time()
    Y = methods[method].fit_transform(df)
    t1 = time()
    time_elapsed = t1-t0
    print("%s: %.2g sec" % (method, time_elapsed))
    ax = fig.add_subplot(3, 2, i + 1)
    ax.scatter(Y[:, 0], Y[:, 1],  cmap=plt.cm.Spectral)
    ax.set_title("%s (%.2g sec)" % (method, time_elapsed))
    ax.xaxis.set_major_formatter(NullFormatter())
    ax.yaxis.set_major_formatter(NullFormatter())
    ax.axis('tight')
  
plt.show()